### PR TITLE
Add extracting static logic into a reducer

### PR DESCRIPTION
**Pull request by ChipmunkForLove**
Step 3 일부까지 완료했습니다.

### CODE REVIEW

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+## reducer로 상태 로직을 통합하기
+
+컴포넌트들의 복잡도가 높아질수록, 컴포넌트들이 업데이트되는 모든 다양한 방식들을 살펴보기는 더욱 더 힘들어집니다. 예를 들어, 아래의 `TaskApp` 컴포넌트는 `tasks` 배열을 상태로 갖고 있고, tasks를 추가하고, 제거하고, 수정하기 위한
+세 개의 다른 이벤트 핸들러를 사용합니다. 
+
+<Code src="https://codesandbox.io/s/yrv1mz?file=/App.js&utm_medium=sandpack"/>
+
+각각의 이벤트 핸들러들은 상태를 업데이트하기 위해 `setTasks`를 부릅니다. 컴포넌트가 커질 수록, 컴포넌트 전체에 뿌려져 있는 상태 로직의 양도 증가합니다. 
+복잡도를 줄이고 모든 로직을 모든 로직을 하나의 접근하기 편한 장소에 유지하기 위해서, **"reducer""라고 불리는 컴포넌트 밖의 하나의 함수에 상태 로직을 옮길 수 있습니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
각각의 이벤트 핸들러들은 상태를 업데이트하기 위해 `setTasks`를 부릅니다. 컴포넌트가 커질수록, 컴포넌트 전체에 뿌려져 있는 상태 로직의 양도 증가합니다. 
복잡도를 줄이고 모든 로직을 하나의 접근하기 편한 장소에 유지하기 위해서, **"reducer""라고 불리는 컴포넌트 밖의 하나의 함수에 상태 로직을 옮길 수 있습니다.
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+Reducer들은 상태를 관리하기 위한 다른 방법입니다. 세 가지 스텝들을 거쳐 `useState`에서 `useReducer`으로 변경할 수 있습니다. 
+
+1. 상태를 설정하는 것에서 액션을 디스패치하는 것으로 **옮기세요.**
+2. 리듀서 함수를 **작성하세요.**
+3. 컴포넌트에서 리듀서를 **사용하세요.**
+
+## 1단계: 상태를 설정하는 것에서 액션을 디스패치하기
+
+현재 있는 이벤트 헨들러들은 상태를 설정해서 무엇을 할지를 명시합니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
현재 있는 이벤트 헨들러들은 상태를 설정해서 *무엇을 할지*를 명시합니다.
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    })
+  );
+}
+
+function handleDeleteTask(taskId) {
+  setTasks(tasks.filter((t) => t.id !== taskId));
+}
+```
+
+로직을 설정하는 모든 관계를 제거하세요. 남은 것은 세 개의 이벤트 핸들러입니다:
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
로직을 설정하는 모든 상태를 제거하세요. 남은 것은 세 개의 이벤트 핸들러입니다:
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+- `handleAddTask(text)`는 사용자가 "Add"를 누르면 호출됩니다.
+- `handleChangeTask(task)`는 사용자가 테스크를 올리거나 "Save"를 누르면 호출됩니다.
+- `handleDeleteTask(taskId)`는 사용자가 "Delete"을 누르면 호출됩니다.
+
+리듀서로 상태를 관리하는 것은 상태를 바로 설정하는 것과는 미묘하게 다릅니다. 상태를 설정해서 React에게 "무엇을 할지" 말하는 것 대신에, 이벤트 핸들러에서 "액션들"을 보내서 "사용자가 방금 무엇을 했는지"를 명시하는 것입니다.
+(로직을 업데이트하는 상태는 어딘가에 살아있을 것입니다!) 그래서 "`테스크들`를 설정하는 것" 대신에 이벤트 핸들러를 통해, "테스크를 추가했고/변경했고/삭제했다"는 행동을 보낼 것입니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
(로직을 업데이트하는 상태는 어딘가에 살아있을 것입니다!) 그래서 이벤트 핸들러를 통해 "`테스크들`를 설정하는 것" 대신에 , "테스크를 추가했고/변경했고/삭제했다"는 행동을 보낼 것입니다. 
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+관례에 따르면, 무엇이 일어났는지를 묘사하는 string `type`을 주는 것이 흔하고, 다른 필드에 어떤 추가적인 정보를 전달 할 수도 있습니다. 
+`type`은 한 컴포넌트에 특정되는 것이기 떄문에 이 경우에는 `added` 혹은 `added_task` 중 하나가 적당해 보입니다. 어떤 일이 일어났는지를 말해주는 이름을 선택하세요! 
+
+```js
+dispatch({
+  // specific to component
+  type: 'what_happened',
+  // other fields go here
+});
+```
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]: 주석 부분도 해석해 주세요!

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+리듀서 함수는 상태 로직을 넣는 위치입니다. 현재의 상태와 action 객체, 두 개의 인자를 받고 다음 상태를 되돌려줍니다:
+
+```js
+function yourReducer(state, action) {
+  // return next state for React to set
+}
+```
+
+React 는 reducer에서 리턴되는 것을 상태로 설정할 것입니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
React 는 reducer에서 반환되는 것을 상태로 설정할 것입니다. 
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+  } else if (action.type === 'deleted') {
+    return tasks.filter((t) => t.id !== action.id);
+  } else {
+    throw Error('Unknown action: ' + action.type);
+  }
+}
+```
+
+<BlockQuote>
+reducer 함수는 상태(task)를 인자로 받기 때문에, 컴포넌트 밖에 그 인자를 선언해야 합니다. 이렇게 하면 들여쓰기 깊이도 줄어들고 코드도 더 쉽게 읽히도록 만들 수 있습니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
reducer 함수는 상태(task)를 인자로 받기 때문에, **컴포넌트 밖에 그 인자를 선언해야 합니다**. 이렇게 하면 들여쓰기 깊이도 줄어들고 코드도 더 쉽게 읽히도록 만들 수 있습니다.
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+```
+각각의 `case`를 `{`와`}`인 중괄호로 감싸는 것을 추천합니다. 다른 `case`의 안에 선언된 변수들이 서로 부딪히지 않도록 하기 위함임니다. 또한, 한 `case`도 보통 `return`으로 끝나야 합니다. 만약 `return`하는 것을 잊는다면, 
+코드는 다음 `case`로 "통과해 떨어져서", 실수로 이어질 수 있습니다!
+
+만약 아직 switch 구문에 익숙하지 않다면 대신 if/else를 사용해도 완전히 무방합니다.
+:::
+
+:::info Deep Dive
+### 왜 reducer는 이러한 방식으로 불리는 것일까요?
+ reducer는 컴포넌트 안ㅔ 코드의 양을 "줄일" 수 있지만, 사실 배열에서 수행 수  있는 reduce() 작업에서 이름지어졌습니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
 reducer는 컴포넌트 안에 코드의 양을 "줄일" 수 있지만, 사실 배열에서 수행할 수  있는 reduce() 작업에서 이름 지어졌습니다. 
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+ ```js
+ const arr = [1, 2, 3, 4, 5];
+const sum = arr.reduce(
+  (result, number) => result + number
+); // 1 + 2 + 3 + 4 + 5
+ ```
+
+`reduce`를 통해 전달한 함수는 "reducer"로 알려져 잇습니다. 이 함수는 이 때까지의 _결과_와 _현재 요소_를 받고나서 _다음 결과_를 리턴해줍니다. React의 reducer 또한 같은 생각의 한 예입니다: 
+_이 때까지의 상태_를 받고 _다음 결과_를 리턴해줍니다. 이러한 방법으로, 상태에 action들을 축적합니다.
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]:
```suggestion
_이 때까지의 상태_를와 _액션_을 받고 _다음 결과_를 리턴해줍니다. 이러한 방법으로, 상태에 action들을 축적합니다.
```

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+심지어 reducer 함수에 넘겨서 최종 상태를 계산하기 위해 `initialState`와 `action`들의 배열로 `reduce()` 메소드를 사용할 수 있습니다:
+
+<Code src="https://codesandbox.io/s/ogkvy1?file=/index.js&utm_medium=sandpack"/>
+<Code src="https://codesandbox.io/s/ogkvy1?file=/tasksReducer.js&utm_medium=sandpack"/>
+<Code src="https://codesandbox.io/s/ogkvy1?file=%2Fpublic%2Findex.html&utm_medium=sandpack"/>
```
Comment (hyehyeonmoon) [inline docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]: sandbox code 링크 하나만 달아도 됩니다!

Comment (ChipmunkForLove) [reply docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]: [f9826f2](https://github.com/LINE-NEXT-LEVEL/client/pull/26/commits/f9826f2e3d46ef9ba5b737fb21c8c40c1975db60)

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx (hunk 7 above)
Comment (ChipmunkForLove) [reply docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]: [bd3f7b4](https://github.com/LINE-NEXT-LEVEL/client/pull/26/commits/bd3f7b4891224024a80efc57e51d085b60eb4f65)

File: docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx (hunk 5 above)
Comment (ChipmunkForLove) [reply docs/learn/Managing State/extracting-state-logic-into-a-reducer.mdx]: [bd3f7b4](https://github.com/LINE-NEXT-LEVEL/client/pull/26/commits/bd3f7b4891224024a80efc57e51d085b60eb4f65)

**Review by imnotmoon — APPROVED** [review]
👍👍👍

**Review by Songycs — APPROVED** [review]
👍 👍 👍

**Pull request merged**